### PR TITLE
WIP Add marathon module

### PR DESCRIPTION
**Pull request by ludovicc**
**ISSUE TYPE**
- New Module Pull Request

**COMPONENT NAME**
- marathon

**ANSIBLE VERSION**
- 2.2

**SUMMARY**

New module to control applications managed by Marathon on a Mesos cluster.

**NOTES**

Starting and stopping applications works well, applications in Docker containers are supported.
Security features of Marathon are not yet supported (SSL, authentication).

### CODE REVIEW

**Comment by sivel** [comment]
Based on the check from travis, this module does not support python 2.4 or python 2.6 due to the use of a dict comprehension.  This will need to be rewritten to support python 2.4 and python 2.6.  (There appear to be no other reasons why this could not support python 2.4)

Additionally, here are some initial points of feedback based on automated checks:

```
============================================================================
marathon.py
============================================================================
ERROR: version_added should be 2.0. Currently 1.9
ERROR: indentation contains tabs. line 302 column 4
ERROR: indentation contains tabs. line 303 column 4
ERROR: indentation contains tabs. line 319 column 4
ERROR: indentation contains tabs. line 320 column 4
ERROR: indentation contains tabs. line 469 column 4
ERROR: indentation contains tabs. line 472 column 4
ERROR: indentation contains tabs. line 475 column 4
ERROR: GPLv3 license header not found
WARNING: No RETURN provided
WARNING: JSON import found, already provided by ansible.module_utils.basic
WARNING: Redeclared basic.py variable or function: json, time
```

**Comment by pegerto** [comment]
Hello,

Do you have an release target for this feature ?

Regards

**Comment by ludovicc** [comment]
@pegerto It depends on the community now, please try this module and tag it 'worksforme'

See REVIEWERS.md:

For inclusion of a new module into Ansible Extras, a pull request must receive at least one approval from a fellow community member on each of the following criteria:

```
One "worksforme" approval from someone who has thoroughly tested the module, including all parameters and switches.
One "passes_guidelines" approval from someone who has vetted the code according to the module guidelines.
```

File: clustering/marathon.py [hunk 1]
```diff
@@ -0,0 +1,555 @@
...
+author: "Ludovic Claude (@ludovicc)"
+
+options:
+  uri:
+    required: true
+    description:
+      - Base URI for the Marathon instance
+
+  state:
+    required: true
```
Comment (resmo) [inline clustering/marathon.py]: remove `required: true` because if you define a default, no input is required.

File: clustering/marathon.py [hunk 2]
```diff
@@ -0,0 +1,555 @@
...
+    choices: [ present, absent, restart, kill ]
+    default: "present"
+    description:
+      - The operation to perform.
+
+  username:
+    required: false
```
Comment (resmo) [inline clustering/marathon.py]: if `required: false` is defined you must define a default e.g. `default: null` also see http://docs.ansible.com/ansible/developing_modules.html#module-checklist

File: clustering/marathon.py [hunk 3]
```diff
@@ -0,0 +1,555 @@
...
+    description:
+      - The username to log-in with.
+
+  password:
+    required: false
```
Comment (resmo) [inline clustering/marathon.py]: if required: false is defined you must define a default e.g. default: null also see http://docs.ansible.com/ansible/developing_modules.html#module-checklist

File: clustering/marathon.py [hunk 4]
```diff
@@ -0,0 +1,555 @@
...
+    description:
+     - The amount of disk in MB that is reserved for the application per instance.
+
+  ports:
+    required: false
+    description:
+     - An array of required port resources on the host.
+
+  requirePorts:
```
Comment (resmo) [inline clustering/marathon.py]: use snake case: require_ports

File: clustering/marathon.py [hunk 5]
```diff
@@ -0,0 +1,555 @@
...
+    required: false
```
Comment (resmo) [inline clustering/marathon.py:104]: if required: false is defined you must define a default e.g. default: null also see http://docs.ansible.com/ansible/developing_modules.html#module-checklist

File: clustering/marathon.py [hunk 6]
```diff
@@ -0,0 +1,555 @@
...
+    description:
+     - Additional data passed to the containerizer on application launch. This is a free-form data structure that can contain arbitrary data.
+
+  docker_image:
+    required: false
+    description:
+     - Name of the Docker image. Ignored if container is defined.
+
+  docker_forcePullImage:
```
Comment (resmo) [inline clustering/marathon.py]: use snake case

File: clustering/marathon.py [hunk 7]
```diff
@@ -0,0 +1,555 @@
...
+    required: false
+    description:
+     - Allows users to run containers in privileged mode. Ignored if container is defined.
+
+  docker_network:
+    required: false
+    description:
+     - Type of networking for the Docker container. Ignored if container is defined.
+
+  docker_portMappings:
```
Comment (resmo) [inline clustering/marathon.py]: please use snake case

File: clustering/marathon.py [hunk 8]
```diff
@@ -0,0 +1,555 @@
...
+    description:
+     - Name of the Docker image. Ignored if container is defined.
+
+  docker_forcePullImage:
+    required: false
+    description:
+     - Force Docker to pull the image before launching each task. Ignored if container is defined.
+
+  docker_privileged:
+    required: false
```
Comment (resmo) [inline clustering/marathon.py:154]: if required: false is defined you must define a default e.g. default: null also see http://docs.ansible.com/ansible/developing_modules.html#module-checklist

File: clustering/marathon.py [hunk 9]
```diff
@@ -0,0 +1,555 @@
...
+    description:
+     - Additional data passed to the containerizer on application launch. This is a free-form data structure that can contain arbitrary data.
+
+  docker_image:
+    required: false
+    description:
+     - Name of the Docker image. Ignored if container is defined.
+
+  docker_forcePullImage:
+    required: false
```
Comment (resmo) [inline clustering/marathon.py:148]: if required: false is defined you must define a default e.g. default: null also see http://docs.ansible.com/ansible/developing_modules.html#module-checklist

File: clustering/marathon.py [hunk 10]
```diff
@@ -0,0 +1,555 @@
...
+    description:
+     - The executor to use to launch this application.
+
+  user:
+    required: false
+    description:
+      - The user to use to launch this application.
+
+  container:
+    required: false
```
Comment (resmo) [inline clustering/marathon.py]: if required: false is defined you must define a default e.g. default: null also see http://docs.ansible.com/ansible/developing_modules.html#module-checklist

File: clustering/marathon.py [hunk 11]
```diff
@@ -0,0 +1,555 @@
...
+    description:
+     - The number of instances of this application to start.
+
+  executor:
+    required: false
+    description:
+     - The executor to use to launch this application.
+
+  user:
+    required: false
```
Comment (resmo) [inline clustering/marathon.py:129]: if required: false is defined you must define a default e.g. default: null also see http://docs.ansible.com/ansible/developing_modules.html#module-checklist

File: clustering/marathon.py [hunk 12]
```diff
@@ -0,0 +1,555 @@
...
+    description:
+     - If true, the ports you have specified are used as host ports.
+
+  instances:
+    required: false
+    description:
+     - The number of instances of this application to start.
+
+  executor:
+    required: false
```
Comment (resmo) [inline clustering/marathon.py:123]: if required: false is defined you must define a default e.g. default: null also see http://docs.ansible.com/ansible/developing_modules.html#module-checklist

File: clustering/marathon.py [hunk 13]
```diff
@@ -0,0 +1,555 @@
...
+    description:
+     - An array of required port resources on the host.
+
+  requirePorts:
+    required: false
+    description:
+     - If true, the ports you have specified are used as host ports.
+
+  instances:
+    required: false
```
Comment (resmo) [inline clustering/marathon.py:117]: if required: false is defined you must define a default e.g. default: null also see http://docs.ansible.com/ansible/developing_modules.html#module-checklist

File: clustering/marathon.py [hunk 14]
```diff
@@ -0,0 +1,555 @@
...
+    description:
+      - The user to use to launch this application.
+
+  container:
+    required: false
+    description:
+     - Additional data passed to the containerizer on application launch. This is a free-form data structure that can contain arbitrary data.
+
+  docker_image:
+    required: false
```
Comment (resmo) [inline clustering/marathon.py:141]: if required: false is defined you must define a default e.g. default: null also see http://docs.ansible.com/ansible/developing_modules.html#module-checklist

File: clustering/marathon.py [hunk 15]
```diff
@@ -0,0 +1,555 @@
...
+    description:
+     - Force Docker to pull the image before launching each task. Ignored if container is defined.
+
+  docker_privileged:
+    required: false
+    description:
+     - Allows users to run containers in privileged mode. Ignored if container is defined.
+
+  docker_network:
+    required: false
```
Comment (resmo) [inline clustering/marathon.py:160]: if required: false is defined you must define a default e.g. default: null also see http://docs.ansible.com/ansible/developing_modules.html#module-checklist

File: clustering/marathon.py [hunk 16]
```diff
@@ -0,0 +1,555 @@
...
+    description:
+     - Type of networking for the Docker container. Ignored if container is defined.
+
+  docker_portMappings:
+    required: false
```
Comment (resmo) [inline clustering/marathon.py:167]: if required: false is defined you must define a default e.g. default: null also see http://docs.ansible.com/ansible/developing_modules.html#module-checklist

File: clustering/marathon.py [hunk 17]
```diff
@@ -0,0 +1,555 @@
...
+    description:
+     - Port mappings for the Docker container. Ignored if container is defined.
+
+  docker_parameters:
+    required: false
```
Comment (resmo) [inline clustering/marathon.py:173]: if required: false is defined you must define a default e.g. default: null also see http://docs.ansible.com/ansible/developing_modules.html#module-checklist

File: clustering/marathon.py [hunk 18]
```diff
@@ -0,0 +1,555 @@
...
+    description:
+     - Arbitrary parameters for the Docker container. Ignored if container is defined.
+
+  container_volumes:
+    required: false
```
Comment (resmo) [inline clustering/marathon.py:185]: if required: false is defined you must define a default e.g. default: null also see http://docs.ansible.com/ansible/developing_modules.html#module-checklist

File: clustering/marathon.py [hunk 19]
```diff
@@ -0,0 +1,555 @@
...
+    description:
+     - Array of volumes for the container. Ignored if container is defined.
+
+  env:
+    required: false
```
Comment (resmo) [inline clustering/marathon.py:197]: if required: false is defined you must define a default e.g. default: null also see http://docs.ansible.com/ansible/developing_modules.html#module-checklist

File: clustering/marathon.py [hunk 20]
```diff
@@ -0,0 +1,555 @@
...
+    description:
+     - Key value pairs that get added to the environment variables of the process to start.
+
+  constraints:
+    required: false
```
Comment (resmo) [inline clustering/marathon.py:203]: if required: false is defined you must define a default e.g. default: null also see http://docs.ansible.com/ansible/developing_modules.html#module-checklist

File: clustering/marathon.py [hunk 21]
```diff
@@ -0,0 +1,555 @@
...
+    description:
+     - Valid constraint operators are one of ["UNIQUE", "CLUSTER", "GROUP_BY"].
+
+  acceptedResourceRoles:
```
Comment (resmo) [inline clustering/marathon.py]: please use snake case

File: clustering/marathon.py [hunk 22]
```diff
@@ -0,0 +1,555 @@
...
+    required: false
```
Comment (resmo) [inline clustering/marathon.py:210]: if required: false is defined you must define a default e.g. default: null also see http://docs.ansible.com/ansible/developing_modules.html#module-checklist

File: clustering/marathon.py [hunk 23]
```diff
@@ -0,0 +1,555 @@
...
+    description:
+     - A list of resource roles.
+
+  labels:
+    required: false
```
Comment (resmo) [inline clustering/marathon.py:216]: if required: false is defined you must define a default e.g. default: null also see http://docs.ansible.com/ansible/developing_modules.html#module-checklist

File: clustering/marathon.py [hunk 24]
```diff
@@ -0,0 +1,555 @@
...
+    description:
+     - Attaching metadata to apps can be useful to expose additional information to other services, so we added the ability to place labels on apps.
+
+  uris:
+    required: false
```
Comment (resmo) [inline clustering/marathon.py:222]: if required: false is defined you must define a default e.g. default: null also see http://docs.ansible.com/ansible/developing_modules.html#module-checklist

File: clustering/marathon.py [hunk 25]
```diff
@@ -0,0 +1,555 @@
...
+    description:
+     - URIs defined here are resolved, before the application gets started. If the application has external dependencies, they should be defined here.
+
+  storeUrls:
+    required: false
```
Comment (resmo) [inline clustering/marathon.py:229]: if required: false is defined you must define a default e.g. default: null also see http://docs.ansible.com/ansible/developing_modules.html#module-checklist

File: clustering/marathon.py [hunk 26]
```diff
@@ -0,0 +1,555 @@
...
+    description:
+     - Attaching metadata to apps can be useful to expose additional information to other services, so we added the ability to place labels on apps.
+
+  uris:
+    required: false
+    description:
+     - URIs defined here are resolved, before the application gets started. If the application has external dependencies, they should be defined here.
+
+  storeUrls:
```
Comment (resmo) [inline clustering/marathon.py]: please use snake case

File: clustering/marathon.py [hunk 27]
```diff
@@ -0,0 +1,555 @@
...
+    required: false
+    description:
+     - a sequence of URIs, that get fetched on each instance, that gets started. The artifact could be fetched directly from the source, or put into the artifact store. One simple way to do this is automatic artifact storing.
+
+  dependencies:
+    required: false
```
Comment (resmo) [inline clustering/marathon.py:235]: if required: false is defined you must define a default e.g. default: null also see http://docs.ansible.com/ansible/developing_modules.html#module-checklist

File: clustering/marathon.py [hunk 28]
```diff
@@ -0,0 +1,555 @@
...
+    description:
+     - A list of services upon which this application depends.
+
+  healthChecks:
```
Comment (resmo) [inline clustering/marathon.py]: snake case please

File: clustering/marathon.py [hunk 29]
```diff
@@ -0,0 +1,555 @@
...
+    required: false
```
Comment (resmo) [inline clustering/marathon.py:248]: if required: false is defined you must define a default e.g. default: null also see http://docs.ansible.com/ansible/developing_modules.html#module-checklist

File: clustering/marathon.py [hunk 30]
```diff
@@ -0,0 +1,555 @@
...
+    description:
+     - An array of checks to be performed on running tasks to determine if they are operating as expected.
+
+  backoffSeconds:
```
Comment (resmo) [inline clustering/marathon.py]: snake case please

File: clustering/marathon.py [hunk 31]
```diff
@@ -0,0 +1,555 @@
...
+    required: false
```
Comment (resmo) [inline clustering/marathon.py:255]: if required: false is defined you must define a default e.g. default: null also see http://docs.ansible.com/ansible/developing_modules.html#module-checklist

File: clustering/marathon.py [hunk 32]
```diff
@@ -0,0 +1,555 @@
...
+    description:
+     - Configures exponential backoff behavior when launching potentially sick apps. The backoff period is multiplied by the factor for each consecutive failure until it reaches maxLaunchDelaySeconds.
+
+  backoffFactor:
+    required: false
+    description:
+     - Configures exponential backoff behavior when launching potentially sick apps. The backoff period is multiplied by the factor for each consecutive failure until it reaches maxLaunchDelaySeconds.
+
+  maxLaunchDelaySeconds:
+    required: false
```
Comment (resmo) [inline clustering/marathon.py:269]: if required: false is defined you must define a default e.g. default: null also see http://docs.ansible.com/ansible/developing_modules.html#module-checklist

File: clustering/marathon.py [hunk 33]
```diff
@@ -0,0 +1,555 @@
...
+    description:
+     - An array of checks to be performed on running tasks to determine if they are operating as expected.
+
+  backoffSeconds:
+    required: false
+    description:
+     - Configures exponential backoff behavior when launching potentially sick apps. The backoff period is multiplied by the factor for each consecutive failure until it reaches maxLaunchDelaySeconds.
+
+  backoffFactor:
+    required: false
```
Comment (resmo) [inline clustering/marathon.py:262]: if required: false is defined you must define a default e.g. default: null also see http://docs.ansible.com/ansible/developing_modules.html#module-checklist

File: clustering/marathon.py [hunk 34]
```diff
@@ -0,0 +1,555 @@
...
+    description:
+     - Configures exponential backoff behavior when launching potentially sick apps. The backoff period is multiplied by the factor for each consecutive failure until it reaches maxLaunchDelaySeconds.
+
+  maxLaunchDelaySeconds:
+    required: false
+    description:
+     - Configures exponential backoff behavior when launching potentially sick apps. The backoff period is multiplied by the factor for each consecutive failure until it reaches maxLaunchDelaySeconds.
+
+  upgradeStrategy_minimumHealthCapacity:
+    required: false
```
Comment (resmo) [inline clustering/marathon.py:276]: if required: false is defined you must define a default e.g. default: null also see http://docs.ansible.com/ansible/developing_modules.html#module-checklist

File: clustering/marathon.py [hunk 35]
```diff
@@ -0,0 +1,555 @@
...
+    description:
+     - a number between 0and 1 that is multiplied with the instance count. This is the minimum number of healthy nodes that do not sacrifice overall application purpose.
+
+  upgradeStrategy_maximumOverCapacity:
+    required: false
```
Comment (resmo) [inline clustering/marathon.py:283]: if required: false is defined you must define a default e.g. default: null also see http://docs.ansible.com/ansible/developing_modules.html#module-checklist

File: clustering/marathon.py [hunk 36]
```diff
@@ -0,0 +1,555 @@
...
+    description:
+     - a number between 0 and 1 which is multiplied with the instance count. This is the maximum number of additional instances launched at any point of time during the upgrade process.
+
+  force:
+    required: false
```
Comment (resmo) [inline clustering/marathon.py:308]: if required: false is defined you must define a default e.g. default: null also see http://docs.ansible.com/ansible/developing_modules.html#module-checklist

File: clustering/marathon.py [hunk 37]
```diff
@@ -0,0 +1,555 @@
...
+    description:
+     - If the app is affected by a running deployment, then the update operation will fail. The current deployment can be overridden by setting the `force` query parameter. Default: false.
+
+  waitTimeout:
+    required: false
```
Comment (resmo) [inline clustering/marathon.py:315]: if required: false is defined you must define a default e.g. default: null also see http://docs.ansible.com/ansible/developing_modules.html#module-checklist

File: clustering/marathon.py [hunk 38]
```diff
@@ -0,0 +1,555 @@
...
+    description:
+     - a number between 0 and 1 which is multiplied with the instance count. This is the maximum number of additional instances launched at any point of time during the upgrade process.
+
+  force:
+    required: false
+    description:
+     - If the app is affected by a running deployment, then the update operation will fail. The current deployment can be overridden by setting the `force` query parameter. Default: false.
+
+  waitTimeout:
```
Comment (resmo) [inline clustering/marathon.py]: use snake case

File: clustering/marathon.py [hunk 39]
```diff
@@ -0,0 +1,555 @@
...
+                   restart=['id'],
+                   kill=['id'])
+
+def main():
+
+    global module
+    module = AnsibleModule(
+        argument_spec=dict(
+            uri=dict(required=True),
+            state=dict(choices=['absent', 'present', 'restart', 'kill'], required=True),
```
Comment (resmo) [inline clustering/marathon.py]: missing default="present", remove required=True.

**Comment by resmo** [comment]
@ludovicc I reviewed it ~20 days ago and made comments, did you have time to look at it? I would like to bring it into the 2.1 release. If you don't have time let me know, so I can make a new PR based on yours with the revisions.

**Comment by ludovicc** [comment]
@resmo thanks very much for your extensive review. I'll try to get the fixes done over this week end

**Comment by gundalow** [comment]
Although not a hard requirement it would be good if the code was PEP8 clean. It's a lot easier for this to be done as part of the initial module creation. If this was to be fixed after the module is submitted it makes tracing where errors got introduced a little harder.

The Makefile in ansible/ansible defines the PEP8 standard to be

`pep8 -r --ignore=E501,E221,W291,W391,E302,E251,E203,W293,E231,E303,E201,E225,E261,E241,E402 marathon.py`

2) Similarly their is a fair bit of trailing whitespace
sed 's/[ \t]*$//' marathon.py

3) You may wish to squash your commits into one

**Comment by jkramarz** [comment]
Hi there; I've implemented few changes suggested here + few fixes in [a pull request to @ludovicc repo](https://github.com/ludovicc/ansible-modules-extras/pull/1) . It works for me, but I'm not happy with `changed=True`.

**Comment by ludovicc** [comment]
@jkramarz changed=True should be no more, thanks

**Comment by gregdek** [comment]
@ludovicc A friendly reminder: this pull request has been marked as needing your action. If you still believe that this PR applies, and you intend to address the issues with this PR, just let us know in the PR itself and we will keep it open pending your changes. When you do address the issues, please respond with ready_for_review in your comment, so that we can notify the maintainer.

[This message brought to you by your friendly Ansibull-bot.]

**Comment by ludovicc** [comment]
ready_for_review

**Comment by gregdek** [comment]
Thanks @ludovicc for this new module. When this module receives 'shipit' comments from two community members and any 'needs_revision' comments have been resolved, we will mark for inclusion.

[This message brought to you by your friendly Ansibull-bot.]

**Comment by jkramarz** [comment]
shipit

File: clustering/marathon.py [hunk 40]
```diff
@@ -0,0 +1,648 @@
...
+    required: false
+    default: 0.0
+    description:
+      - The amount of disk in MB that is reserved for the application per instance.
+
+  ports:
+    required: false
+    default: '[]'
+    description:
+      - An array of required port resources on the host.
```
Comment (gundalow) [inline clustering/marathon.py]: Does this need the following adding:

An array of required port resources on the host. Required if C(require_ports=true).

File: clustering/marathon.py [hunk 41]
```diff
@@ -0,0 +1,648 @@
...
+    default: 3600
+    description:
+      - Configures exponential backoff behavior when launching potentially sick apps. The backoff period is multiplied by the factor for each consecutive failure until it reaches maxLaunchDelaySeconds.
+
+  upgrade_strategy_minimum_health_capacity:
+    aliases: [ upgradeStrategy_minimumHealthCapacity ]
+    required: false
+    default: 1.0
+    description:
+      - a number between 0 and 1 that is multiplied with the instance count. This is the minimum number of healthy nodes that do not sacrifice overall application purpose.
```
Comment (gundalow) [inline clustering/marathon.py]: "A number..."

File: clustering/marathon.py [hunk 42]
```diff
@@ -0,0 +1,648 @@
...
+
+  upgrade_strategy_maximum_over_capacity:
+    aliases: [ upgradeStrategy_maximumOverCapacity ]
+    required: false
+    default: 0.0
+    description:
+      - a number between 0 and 1 which is multiplied with the instance count. This is the maximum number of additional instances launched at any point of time during the upgrade process.
```
Comment (gundalow) [inline clustering/marathon.py]: A Number...

File: clustering/marathon.py (hunk 40 above)
Comment (ludovicc) [inline clustering/marathon.py]: @gundalow indeed, thanks

**Comment by ludovicc** [comment]
@pegerto @gundalow @jkramarz please test and mark this plugin as ready for shipping, let's release this for Ansible 2.1!

**Comment by ludovicc** [comment]
I rebased and squashed the commits

File: clustering/marathon.py [hunk 43]
```diff
@@ -0,0 +1,648 @@
...
+def main():
+
+    global module
+    module = AnsibleModule(
+        argument_spec=dict(
+            uri=dict(required=True),
+            state=dict(default='present', choices=['absent', 'present', 'restart', 'kill']),
+            username=dict(required=False,default=None),
+            password=dict(required=False,default=None),
+            id=dict(type='str'),
```
Comment (resmo) [inline clustering/marathon.py]: according to the docs, id is required, is it?

Comment (ludovicc) [inline clustering/marathon.py]: @resmo yes it is, thanks

File: clustering/marathon.py [hunk 44]
```diff
@@ -0,0 +1,686 @@
...
+    except Exception, e:
+        return module.fail_json(msg=e.message)
+
+
+    module.exit_json(changed=ret.changed, uri=uri, state=state, meta=ret.meta)
+
+
+from ansible.module_utils.basic import *
+from ansible.module_utils.urls import *
+main()
```
Comment (gundalow) [inline clustering/marathon.py]: Could you please update this to:

```
if __name__ == '__main__':
    main()
```

As per http://docs.ansible.com/ansible/developing_modules.html#common-module-boilerplate

**Comment by gundalow** [comment]
needs_revision

File: clustering/marathon.py (hunk 44 above)
Comment (ludovicc) [inline clustering/marathon.py]: @gundalow ok, done

**Comment by gundalow** [comment]
I'm giving this a shipit (though I don't marathon setup to be able to test this).

**Comment by ludovicc** [comment]
Tested with Marathon 1.1.1. Please shipit. To test with Marathon, all you need is docker-compose and https://github.com/meltwater/docker-mesos

**Comment by ludovicc** [comment]
@gregdek new plugin ready for inclusion, tested in production, please review

**Comment by resmo** [comment]
@ludovicc thanks for the update, I looked over the code and to me it is not clear if the module does handle "idempotence". 

Is it possible to run the same task twice and the second run will show a `changed: false` and will not do any work? To me this is essential.

Otherwise it would just be a wrapper to curl without much benefit. 
Thanks for clarification.

**Comment by gregdek** [comment]
Thanks again to @ludovicc for this PR, and thanks @ansible/core for reviewing. Marking for inclusion.

[This message brought to you by your friendly Ansibull-bot.]

**Comment by ludovicc** [comment]
@resmo this module is already more than a wrapper to curl: it contains some logic to wait for Marathon to complete its deployment, it simplifies the configuration to launch Docker containers, it validates some parameters and handles type conversions so you can parameterise some integer or float value using the syntax "{{ my_param }}". 

I agree that idempotence is a required feature in an Ansible module, and I will try to implement that also.

**Comment by alikins** [comment]
Seems to be some confusion about the auto-shipit  (and as of today, if 'shipit' is a thing anymore)

**Comment by gregdek** [comment]
Thanks again to @ludovicc for this PR, and thanks @ansible/core for reviewing. Marking for inclusion.

[This message brought to you by your friendly Ansibull-bot.]

**Comment by resmo** [comment]
Thanks @ludovicc for clarification. Since this module won't make it into 2.1, we have plenty of time to make it idempotent for 2.2 and to get it merged in. 

@gundalow gave it a ship it so this is marked for inclusion. 

I am not merging as is as impotence is a requirement for me, IMHO it is the first thing users will complain and as I saw marathon has a REST API it is doable.

Leaving if for someone else from ansible org, who wants to merge it in as is.

**Comment by nitzmahone** [comment]
Agreed: let's get idempotence going.

**Comment by gregdek** [comment]
Thanks @ludovicc for this new module. When this module receives 'shipit' comments from two community members and any 'needs_revision' comments have been resolved, we will mark for inclusion.

[This message brought to you by your friendly Ansibull-bot.]

**Comment by gundalow** [comment]
needs_revision

Can you please add idempotence as stated by @resmo and @nitzmahone

**Comment by gregdek** [comment]
Thanks @ludovicc for this PR. This PR requires revisions, either because it fails to build or by reviewer request. Please make the suggested revisions. When you are done, please comment with text 'ready_for_review' and we will put this PR back into review.

[This message brought to you by your friendly Ansibull-bot.]

**Comment by gregdek** [comment]
@ludovicc A friendly reminder: this pull request has been marked as needing your action. If you still believe that this PR applies, and you intend to address the issues with this PR, just let us know in the PR itself and we will keep it open pending your changes. When you do address the issues, please respond with ready_for_review in your comment, so that we can notify the maintainer.

[This message brought to you by your friendly Ansibull-bot.]

**Comment by gregdek** [comment]
@ludovicc Another friendly reminder: this pull request has been marked as needing your action. If you still believe that this PR applies, and you intend to address the issues with this PR, just let us know in the PR itself and we will keep it open. If you have addressed the issues and believe it's ready for review, please comment with the text "ready_for_review". If we don't hear from you within another 14 days, we will close this pull request.

[This message brought to you by your friendly Ansibull-bot.]

**Comment by gregdek** [comment]
@ludovicc A friendly reminder: this pull request has been marked as needing your action. If you still believe that this PR applies, and you intend to address the issues with this PR, just let us know in the PR itself and we will keep it open pending your changes. When you do address the issues, please respond with ready_for_review in your comment, so that we can notify the maintainer.

[This message brought to you by your friendly Ansibull-bot.]

**Comment by ludovicc** [comment]
I still need to find some time to work on this issue.

2016-07-04 12:29 GMT+02:00 Greg DeKoenigsberg notifications@github.com:

**Comment by gundalow** [comment]
@ludovicc I think you you update the title of this PR to start with "WIP" then the bot will ignore it

WIP = Work In Progress

**Comment by gregdek** [comment]
Thanks @ludovicc for this new module. When this module receives 'shipit' comments from two community members and any 'needs_revision' comments have been resolved, we will mark for inclusion.

[This message brought to you by your friendly Ansibull-bot.]

**Comment by gregdek** [comment]
This PR is a work-in-progress. @ludovicc, when you're ready to have this PR reviewed, please remove WIP from the title of the PR.

**Pull request closed without merging**